test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
+}
+
+func TestPodOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-pod-orphaning-and-adoption-when-labels-change", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 1)
+	}
+
+	pod := &pods.Items[0]
+
+	// Start by verifying ControllerRef for the pod is not nil
+	if metav1.GetControllerOf(pod) == nil {
+		t.Fatalf("ControllerRef of pod %s is nil", pod.Name)
+	}
+
+	newLabelMap := map[string]string{"new-foo": "new-bar"}
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPod.Labels = newLabelMap
+		_, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update labels for pod %s: %v", pod.Name, err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Then, verify ControllerRef for the pod is nil
+		if metav1.GetControllerOf(newPod) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify ControllerRef for the pod %s is nil: %v", pod.Name, err)
+	}
+
+	// Adoption: RS should add ControllerRef to a pod when the pod's labels change to match its labels
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Revert to original labels so that the RS can adopt the pod again
+		newPod.Labels = labelMap
+		_, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to revert to original labels for pod %s: %v", pod.Name, err)
+	}
+
+	controllerRef := metav1.GetControllerOf(pod)
```

Comment:
I was aware that there might be a need to fetch pod here using `GET` after its labels were reverted. As its controllerRef UID matches the RS's UID (see below), it supported my understanding that reassignment `pod = newPod` might not be necessary.

Reply:
Above, you updated the labels. We then expect RS to take action asynchronously. Therefore we need to poll to check for the result of that action.

Your current code only appears to pass because it uses an old value of `pod` from before it was orphaned.

Reply:
I was confused about necessity of `pod = newPod`, but I understand why it is needed now after getting back to this PR with fresh mind. To clarify, value of `pod` has to be updated (or fetched again) every time a change happens (in this test, it is orphaning or adoption). Otherwise, the test will compare the controller UIDs like orphaning and adoption does not happen at all.